dataset/bins.cpp

```diff
@@ -405,4 +405,53 @@ Dataset sum(const Dataset &d) {
   return apply_to_items(d, [](auto &&... _) { return buckets::sum(_...); });
 }
 
+Variable mean(const Variable &data) {
+  auto type = variable::variableFactory().elem_dtype(data);
+  type = type == dtype<bool> ? dtype<int64_t> : type;
+  const auto unit = variable::variableFactory().elem_unit(data);
+
+  Variable means;
+  if (variable::variableFactory().hasVariances(data))
+    means = Variable(type, data.dims(), unit, Values{}, Variances{});
+  else
+    means = Variable(type, data.dims(), unit, Values{});
+
+  if (data.dtype() == dtype<bucket<DataArray>>) {
+    const auto &&[indices, dim, buffer] = data.constituents<DataArray>();
+    if (const auto mask_union = irreducible_mask(buffer.masks(), dim);
+        mask_union.is_valid()) {
+      const auto masked = applyMask(buffer, indices, dim, mask_union);
+      variable::sum_impl(means, masked);
+
+      // Trick to get the sizes of bins if masks are present - bin the masks
+      // using the same dimension & indices as the data, and then sum the
+      // inverse of the mask to get the number of unmasked entries.
+      auto sizes = makeVariable<int64_t>(masked.dims());
+      variable::sum_impl(sizes, make_bins(indices, dim, ~mask_union));
+
+      means /= astype(sizes, dtype<double>);
+
+    } else {
+      variable::sum_impl(means, data);
+      means = astype(means, dtype<double>, CopyPolicy::TryAvoid) /
+              astype(bucket_sizes(data), dtype<double>, CopyPolicy::TryAvoid);
+    }
+  } else {
+    variable::sum_impl(means, data);
+    means = astype(means, dtype<double>, CopyPolicy::TryAvoid) /
+            astype(bucket_sizes(data), dtype<double>, CopyPolicy::TryAvoid);
```

Comment:
Could this just use `sum` from above? Even before creating `means`, just
```cpp
return sum(data) / bucket_sizes(data);
```
or something like that?

Also, you have an issue with the dtype selection here, always returning `double`, even if input was, e.g., `float`?